feat(wallet): Derive beyond 10 ledger accounts when importing

marc2332 opened the pull request:
Closes https://github.com/iotaledger/iota/issues/9285

marc2332 committed:
feat(wallet): Derive beyond 10 ledger accounts when importing

marc2332 committed:
Merge branch 'develop' into tooling/derive-beyong-10-ledger-accounts-when-importing

begonaalvarezd committed:
Merge branch 'develop' into tooling/derive-beyong-10-ledger-accounts-when-importing

begonaalvarezd committed:
feat(wallet): polish ui

VmMad commented on apps/wallet/src/ui/app/pages/accounts/ImportLedgerAccountsPage.tsx:
remove this pls

VmMad reviewed (changes requested):
There's a couple things to improve:
1. Now the first account in the ledger doesn't show up anymore:
(DEV)
<img width="370" height="701" alt="Image" src="https://github.com/user-attachments/assets/084e5893-ae4c-4375-8ceb-2f0f7c8ea971" />

(RC)
<img width="370" height="701" alt="Image" src="https://github.com/user-attachments/assets/ad3e4ad4-ce11-451f-9de0-ebdd22a975f5" />

When first importing the accounts in a new ledger this screen appears for a second: 

<img width="756" height="992" alt="Image" src="https://github.com/user-attachments/assets/1237d178-0680-4977-b4e4-2b1db84dfd34" />

then it starts loading. Also sometimes that screen is shown even when the accounts are not imported at all. I could trigger it by doing basically:
1. Try to import the ledger
2. In the account selection just click the top right close.
3. Try to import the ledger again
4. The scrren shows a fixed message `Imported all Ledger Accounts` but that's fake info

marc2332 committed:
fix starting deriving index

marc2332 committed:
Fixes

marc2332 committed:
;

marc2332 committed:
Merge branch 'develop' into tooling/derive-beyong-10-ledger-accounts-when-importing

marc2332 committed:
improvements

VmMad commented on apps/wallet/src/ui/app/pages/accounts/ImportLedgerAccountsPage.tsx:
```diff
@@ -15,35 +15,34 @@ import {
+            error: ledgerError,
+            isPending: areLedgerAccountsLoading,
+            isError: encounteredDerviceAccountsError,
+            mutateAsync: loadMore,
         },
```
```suggestion
            mutateAsync: loadMore,
            isIdle: isIdleDeriveAccounts,
},
```

VmMad commented on apps/wallet/src/ui/app/pages/accounts/ImportLedgerAccountsPage.tsx:
```diff
@@ -133,22 +135,30 @@ export function ImportLedgerAccountsPage() {
+                <div className="flex flex-1 items-end gap-xs">
+                    <Button
+                        type={ButtonType.Secondary}
+                        disabled={areLedgerAccountsLoading}
```
```suggestion
                        disabled={isLoading}
```

marc2332 replied:
`isLoading` is better

marc2332 committed:
Update apps/wallet/src/ui/app/pages/accounts/ImportLedgerAccountsPage.tsx

Co-authored-by: JCNoguera <88061365+VmMad@users.noreply.github.com>

KeitoTadashi commented on apps/wallet/src/ui/app/components/ledger/useDeriveLedgerAccounts.ts:
```diff
@@ -5,49 +5,70 @@
-        'select'
-    > {
-    numAccountsToDerive: number;
+interface UseDeriveLedgerAccountOptions {
```
```suggestion
interface UseDeriveLedgerAccountOptions {
    chunkSize: number;
    existingAccounts?: SerializedUIAccount[];
}
```

KeitoTadashi commented on apps/wallet/src/ui/app/components/ledger/useDeriveLedgerAccounts.ts:
```diff
@@ -5,49 +5,70 @@
 
-export function useDeriveLedgerAccounts(options: UseDeriveLedgerAccountOptions) {
-    const { numAccountsToDerive, ...useQueryOptions } = options;
+export function useDeriveLedgerAccounts({ chunkSize }: UseDeriveLedgerAccountOptions) {
```
```suggestion
export function useDeriveLedgerAccounts({
    chunkSize,
    existingAccounts,
}: UseDeriveLedgerAccountOptions) {
```

KeitoTadashi commented on apps/wallet/src/ui/app/components/ledger/useDeriveLedgerAccounts.ts:
```diff
@@ -5,49 +5,70 @@
     const { iotaLedgerClient } = useIotaLedgerClient();
+    const [accounts, setAccounts] = useState<DerivedLedgerAccount[]>([]);
+    const [chunk, setChunk] = useState(0);
 
```
```suggestion
    const [numAccountsToSkip, setNumAccountsToSkip] = useState(0);

```

KeitoTadashi commented on apps/wallet/src/ui/app/components/ledger/useDeriveLedgerAccounts.ts:
```diff
@@ -5,49 +5,70 @@
                 throw new Error("The IOTA application isn't open on a connected Ledger device");
             }
-            return deriveAccountsFromLedger(iotaLedgerClient, numAccountsToDerive);
+            const data = await deriveMainAccountFromLedger(iotaLedgerClient);
```
```suggestion
            const data = await deriveMainAccountFromLedger(iotaLedgerClient);

            const numAlreadyImported = existingAccounts
                ? existingAccounts.filter(
                      (acc) => isLedgerAccountSerializedUI(acc) && acc.mainPublicKey === data,
                  ).length
                : 0;
            setNumAccountsToSkip(numAlreadyImported);

            await advance.mutateAsync(numAlreadyImported);
```

KeitoTadashi commented on apps/wallet/src/ui/app/components/ledger/useDeriveLedgerAccounts.ts:
```diff
@@ -5,49 +5,70 @@
-        ...useQueryOptions,
         gcTime: 0,
+        staleTime: 0,
+        enabled: !!iotaLedgerClient,
```
```suggestion
        enabled: !!iotaLedgerClient && !!existingAccounts,
```

KeitoTadashi commented on apps/wallet/src/ui/app/components/ledger/useDeriveLedgerAccounts.ts:
```diff
@@ -5,49 +5,70 @@
     });
+
+    const advance = useMutation({
+        mutationFn: async () => {
```
```suggestion
        mutationFn: async (startIndexOverride?: number) => {

```

KeitoTadashi commented on apps/wallet/src/ui/app/components/ledger/useDeriveLedgerAccounts.ts:
```diff
@@ -5,49 +5,70 @@
+                throw new Error("The IOTA application isn't open on a connected Ledger device");
+            }
+
+            const newAccounts = await deriveAccountsFromLedger(iotaLedgerClient, chunk, chunkSize);
```
```suggestion
            const startIndex = startIndexOverride ?? numAccountsToSkip + accounts.length;
            const newAccounts = await deriveAccountsFromLedger(
                iotaLedgerClient,
                startIndex,
                chunkSize,
            );
```

KeitoTadashi commented on apps/wallet/src/ui/app/components/ledger/useDeriveLedgerAccounts.ts:
```diff
@@ -5,49 +5,70 @@
+            const newAccounts = await deriveAccountsFromLedger(iotaLedgerClient, chunk, chunkSize);
+
+            setAccounts((accs) => [...accs, ...newAccounts]);
+            setChunk((c) => c + 1);
```
```suggestion
```

KeitoTadashi commented on apps/wallet/src/ui/app/components/ledger/useDeriveLedgerAccounts.ts:
```diff
@@ -5,49 +5,70 @@
-
-    const mainPublicKeyResult = await iotaLedgerClient.getPublicKey(`m/44'/4218'/0'/0'/0'`);
-    const mainPublicKey = new Ed25519PublicKey(mainPublicKeyResult.publicKey);
+    const derivationPaths = getDerivationPathsForLedger(chunk * chunkSize, chunkSize);
```
```suggestion
    const derivationPaths = getDerivationPathsForLedger(chunk, chunkSize);

```

KeitoTadashi commented on apps/wallet/src/ui/app/pages/accounts/ImportLedgerAccountsPage.tsx:
```diff
@@ -15,35 +15,34 @@ import {
+            isError: encounteredDerviceAccountsError,
+            mutateAsync: loadMore,
         },
+    } = useDeriveLedgerAccounts({
+        chunkSize: LEDGER_ACCOUNTS_DERIVE_CHUNKS_SIZE,
     });
 
```
```suggestion
    } = useDeriveLedgerAccounts({
        chunkSize: LEDGER_ACCOUNTS_DERIVE_CHUNKS_SIZE,
        existingAccounts,
    });

```

KeitoTadashi reviewed (changes requested):
LGTM and it works, but:

When reopening the Ledger import screen, accounts are derived starting from index 0 in chunks of 10. Already-imported accounts are filtered out from the displayed list. 

e.g:  n* 10 accounts were previously imported, the first n chunks yield 0 new accounts, so the user sees "All accounts already imported" and must click in "Load More" n times before any unimported accounts appear

Also probably we will want to remove "All accounts already imported"

marc2332 committed:
Update "all accounts imported" message

KeitoTadashi committed:
chore: a test

chore: fix index

chore: unused

chore: add isDisable, and revert filtering

chore: reverts

fix: fmt

marc2332 committed:
.

marc2332 committed:
clean up

marc2332 committed:
changeset

marc2332 committed:
Merge branch 'develop' into tooling/derive-beyong-10-ledger-accounts-when-importing

VmMad reviewed (changes requested):
would be nice to be able to show the already imported accounts as selected (even though they are disabled) and also change the header checkbox to not modify those elements that are disabled. other than that looks nice

marc2332 committed:
only check those accounts not-imported

marc2332 committed:
Merge branch 'tooling/derive-beyong-10-ledger-accounts-when-importing' of https://github.com/iotaledger/iota into tooling/derive-beyong-10-ledger-accounts-when-importing

marc2332 committed:
Merge branch 'develop' into tooling/derive-beyong-10-ledger-accounts-when-importing

marc2332 committed:
Merge branch 'develop' into tooling/derive-beyong-10-ledger-accounts-when-importing

marc2332 committed:
fix

VmMad reviewed (approved):
noice 🆒

VmMad committed:
Merge branch 'develop' into tooling/derive-beyong-10-ledger-accounts-when-importing

marc2332 committed:
Merge branch 'develop' into tooling/derive-beyong-10-ledger-accounts-when-importing

marc2332 committed:
prettier

msarcev committed:
Merge branch 'develop' into tooling/derive-beyong-10-ledger-accounts-when-importing

marc2332 committed:
Fix account click handler to use correct account

marc2332 committed:
fix
